src/common/snippets/src/pass/common_optimizations.cpp

```diff
@@ -68,82 +82,87 @@ void CommonOptimizations::SplitDimensionM(const std::shared_ptr<ov::snippets::op
     // [113745] Heurestic is equal to double block size.
     // When this optimization will be moved into Subgraph and blocking param will be implemented as dependents of shapes,
     // need to implement common way (for all backends) to calculate optimal value of M dimension
     const auto optimal_m_dim = 32 * 2;
-    const auto optimal_parallelism_work_amount = minimal_concurrency;
-    if (m_dim <= optimal_m_dim)
+    const auto optimal_parallelism_work_amount = concurrency;
+    // We skip optimization
+    //  - if the current M dimension is not enough for splitting
+    //  - if the current batch is optimal for concurrency
+    if (m_dim <= optimal_m_dim || batch_dim % optimal_parallelism_work_amount == 0)
         return;
 
-    const auto batch_dim =
-        std::accumulate(mm_shape.rbegin() + 2, mm_shape.rend(), size_t(1), std::multiplies<size_t>());  // B (batch)
     size_t batch_m_dim = 1;
     size_t new_m_dim = m_dim;
 
+    auto is_optimized = [&](size_t batch_m_dim, size_t new_m_dim) {
+        return batch_m_dim > 1 && new_m_dim >= static_cast<size_t>(optimal_m_dim);
+    };
+
+    // [ First Step ]
     // Need to find optimized dimension splitting: [b1..bk, m, n] -> [b1..bk, batch_m_dim, new_m_dim, n]
     // The work amount for parallelism should be divided by max thread count in ideal case
     // that all threads have the same full work amount (avoid of thread downtime)
-    // If it's impossible, it should be more than max thread count
-    // [115284]: Find solution for finding of optimal splitting in these cases
+    // If it's impossible, we select such values so that as many threads as possible have work (see [ Second Step ])
     // For example, there are 16 threads and shape [6, 512, 32]
     //              LCM(6, 16) = 48 <- ideal work amount for parallelism
     //              new_shape [6, 48 / 6, 512 / (48 / 6), 32 ] => [6, 8, 64, 32]
     //              Each thread has parallelism_work_amount = 6 * 8 / nthrs = 3
-    auto get_lcm = [](size_t a, size_t b) {
-        std::function<size_t(size_t, size_t)> get_gcd;
-        get_gcd = [&get_gcd](size_t a, size_t b) {
-            if (b == 0)
-                return a;
-            return get_gcd(b, a % b);
-        };
-        return a / get_gcd(a, b) * b;
-    };
     const auto lcm = get_lcm(batch_dim, optimal_parallelism_work_amount);  // LCM(b, nthrs)
     const auto batch_dim_multiplier = lcm / batch_dim;  // LCM(b, nthrs) / b
     const auto needed_new_dim = m_dim / batch_dim_multiplier;  // m / (LCM(b, nthrs) / b) - needed factors of dimension m
-
-    auto is_optimized = [&](size_t batch_m_dim, size_t new_m_dim) {
-        return batch_m_dim != 1 && new_m_dim >= static_cast<size_t>(optimal_m_dim);
-    };
-
     if (batch_dim_multiplier * needed_new_dim == m_dim) {
         batch_m_dim = batch_dim_multiplier;
         new_m_dim = needed_new_dim;
```

Comment:
Minor: it seems a bit confusing that the Steps are not mutually exclusive: even if we performed decomposition on the first step, we might still fall-through and attempt to decompose again in the second step.
Can we just add smth like `&& batch_dim_multiplier != 1` to the first `if` and replace the second `if` with `else if`?

Reply:
Honestly I don't understand why the steps are not exclusive.

Let's take a look at the example:

We try to optimize shapes at first step. If we find `LCM` and `m_dim % batch_dim_multiplier == 0`, we update `batch_m_dim` and `new_m_dim` here. 
Then we go to the line 134 `if (!is_optimized(batch_m_dim, new_m_dim))`. This check is just to verify that first step has already optimized. If not we try do it again using second step.

Just to remind: `is_optimized`: `return batch_m_dim > 1 && new_m_dim >= static_cast<size_t>(optimal_m_dim);`

There may be the following cases:
- We find `LCM` but unfortunately  `m_dim % batch_dim_multiplier != 0` - we don't update `batch_m_dim` and `new_m_dim` (they save previous values `1` and `m_dim`). So `is_optimized` returns `false` and we try second step. It's OK;
- We find `LCM` and  `m_dim % batch_dim_multiplier == 0`. Great! Let's think that `batch_dim_multiplier  = 1` and `new_m_dim = m_dim`  (it's possible). Anyway `is_optimized` returns `false` since `batch_m_dim = batch_dim_multiplier = 1` not `> 1`. Well, go to the second step.
- Like the previous case but `batch_dim_multiplier  = m_dim` and `new_m_dim = 1`  (it's possible as well). Anyway `is_optimized` returns `false` since `new_m_dim =  1` not `> optimal_m_dim`. Well, got the second step.
- The great case when We find `LCM`,  `m_dim % batch_dim_multiplier == 0` and `is_optimized` returns `true`. We skip second step.

I hope it's clearer now. I just show that if we optimized shapes at first step, we skip second step. Or maybe I didn't understand you point. In this case could you please eloborate? Thank you for the comment!

Reply:
> I hope it's clearer now. I just show that if we optimized shapes at first step, we skip second step.
Indeed! So my suggestion is just to make it more obvious by using `else if`. Otherwise you have to dig into the logic of these checks, and it reduces readability.

Reply:
Changed. I really hope it's obvious now 🙂 Thanks!